presto-main/src/main/java/io/prestosql/sql/planner/DesugarAtTimeZoneRewriter.java

```diff
@@ -55,32 +53,44 @@ public static Expression rewrite(Expression expression, Session session, Metadat
         }
         Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression);
 
-        return rewrite(expression, expressionTypes);
+        return rewrite(expression, expressionTypes, metadata);
     }
 
     private static class Visitor
             extends ExpressionRewriter<Void>
     {
         private final Map<NodeRef<Expression>, Type> expressionTypes;
+        private final Metadata metadata;
 
-        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes)
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
         {
             this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
         }
 
         @Override
         public Expression rewriteAtTimeZone(AtTimeZone node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
+            Type valueType = getType(node.getValue());
             Expression value = treeRewriter.rewrite(node.getValue(), context);
-            Type type = getType(node.getValue());
-            if (type.equals(TIME)) {
+
+            if (valueType.equals(TIME)) {
+                valueType = TIME_WITH_TIME_ZONE;
                 value = new Cast(value, TIME_WITH_TIME_ZONE.getDisplayName());
             }
-            else if (type.equals(TIMESTAMP)) {
+            else if (valueType.equals(TIMESTAMP)) {
+                valueType = TIMESTAMP_WITH_TIME_ZONE;
                 value = new Cast(value, TIMESTAMP_WITH_TIME_ZONE.getDisplayName());
             }
 
-            return new FunctionCall(QualifiedName.of("at_timezone"), ImmutableList.of(value, treeRewriter.rewrite(node.getTimeZone(), context)));
+            Type timeZoneType = getType(node.getTimeZone());
+            Expression timeZone = treeRewriter.rewrite(node.getTimeZone(), context);
+
+            return new FunctionCallBuilder(metadata)
+                    .setName(QualifiedName.of("at_timezone"))
```

Comment:
This `QualifiedName.of()` everywhere looks ugly. How about adding a variant that takes `String`?